### PR TITLE
Solver fixes and API improvements

### PR DESCRIPTION
**Pull request by threecgreen**
* Improved constructors
  * Add ability to specify `stopping_state` and `step_callback` in the constructors of `Solver` and all child classes
  * `Lbfgsb` constructor takes `lower_bound` and `upper_bound` instead of using current hard-coded values. Updated simple and verify code to specify bounds when testing `Lbfgsb`
* Fix build issue with clang from using private member `Dim` of `Function` in `solver.h:175`
* Fix f delta violation count bug where `f_delta_violations` was always reset to 0
* Minor changes:
  * Add additional IDE files to `.gitignore`
  * Modernize linesearch namespace declarations for C++17

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+Language:        Cpp
+BasedOnStyle:  Google
```
Comment (PatWie) [inline .clang-format:3]: Please remove that change from this PR. Could be a separate PR.

File: include/cppoptlib/solver/bfgs.h [hunk 2]
```diff
@@ -22,6 +24,12 @@ class Bfgs : public Solver<function_t> {
  public:
   EIGEN_MAKE_ALIGNED_OPERATOR_NEW
 
+  explicit Bfgs(const State<scalar_t> &stopping_state =
```
Comment (PatWie) [inline include/cppoptlib/solver/bfgs.h:27]: Can't this ctor be part of the parent class?

File: include/cppoptlib/solver/lbfgsb.h [hunk 3]
```diff
@@ -46,21 +55,15 @@ class Lbfgsb : public Solver<function_t> {
   function_state_t OptimizationStep(const function_t &function,
                                     const function_state_t &current,
                                     const state_t & /*state*/) override {
-    const vector_t upper_bound =
```
Comment (PatWie) [inline include/cppoptlib/solver/lbfgsb.h:49]: Yes. Much better!

File: include/cppoptlib/solver/lbfgsb.h [hunk 4]
```diff
@@ -113,39 +116,39 @@ class Lbfgsb : public Solver<function_t> {
...
     // TODO(patwie): use "std::set" ?
     std::vector<std::pair<int, scalar_t>> set_of_t;
+    set_of_t.reserve(dim_);
     // The feasible set is implicitly given by "set_of_t - {t_i==0}".
     vector_t d = -current.gradient;
     // n operations
     for (int j = 0; j < dim_; j++) {
       if (current.gradient(j) == 0) {
-        set_of_t.push_back(std::make_pair(j, max_value));
+        set_of_t.emplace_back(j, max_value);
```
Comment (PatWie) [inline include/cppoptlib/solver/lbfgsb.h:143]: Good catch!

File: include/cppoptlib/solver/lbfgsb.h [hunk 5]
```diff
@@ -302,6 +302,8 @@ class Lbfgsb : public Solver<function_t> {
 
  private:
   int dim_;
+  vector_t lower_bound_;
```
Comment (PatWie) [inline include/cppoptlib/solver/lbfgsb.h:305]: We had a default value before (-inf, info). We should initialize these bound again by default if no other bound is given

**Review by PatWie — CHANGES_REQUESTED** [review]
Could you please have another look at my comments especially about the initialization of the bound.

File: .clang-format (hunk 1 above)
Comment (threecgreen) [reply .clang-format]: removed

File: include/cppoptlib/solver/bfgs.h (hunk 2 above)
Comment (threecgreen) [reply include/cppoptlib/solver/bfgs.h:27]: It is defined for `Solver`, but constructors aren't inherited. [Per cppreference](https://en.cppreference.com/w/cpp/language/using_declaration#Inheriting_constructors), I could expose the `Solver` constructor in the child classes by adding
```cpp
using Superclass::Superclass;
```
but I don't find this very readable or that saves much code. What do you think?

File: include/cppoptlib/solver/lbfgsb.h (hunk 5 above)
Comment (threecgreen) [reply include/cppoptlib/solver/lbfgsb.h:305]: Do you mean set a default in `InitializeSolver`? For functions where `Dim = Eigen::Dynamic`, I'll need `dim_` to be initialized to set default bounds of the correct size.

The way I have it now, `Lbfgsb` has to be initialized with both bounds. Why would someone use a bounded solver if they don't want to impose any bounds?

**Comment by threecgreen** [comment]
Thank you for writing this library and merging my pull request. 

I think these changes should resolve both outstanding issues: #131 and #114.

**Comment by PatWie** [comment]
Agree. But I don't see the link to #114  which talks about line search methods that are aware of bounds. This hasn't changed so far.
